### PR TITLE
Add config file

### PR DESCRIPTION
**Pull request by patxoca**
See issue #16

I've tried to make simple to manage the options while keeping the code minimal. A nice improvement would be creating the config parser dynamically from the schema but I felt like being too much.

I've done some profiling and looking for, reading an parsing the config file does not hurt performance noticeably (milliseconds in an old i5 laptop).

The code requires testing on windows.

### CODE REVIEW

File: z80count/z80count.py [hunk 1]
```diff
@@ -34,6 +37,180 @@
...
+    return args
+
+
+def load_config_file(config_file, schema):
+    parser = configparser.ConfigParser()
+    parser["z80count"] = {i.config_name: i.default for i in schema}
+    try:
+        parser.read(config_file)
+    except configparser.ParsingError:
+        perror("Error parsing config file. Using defaults.")
```
Comment (reidrac) [inline z80count/z80count.py]: May be provide some pointer on what when wrong?

File: z80count/z80count.py [hunk 2]
```diff
@@ -34,6 +37,180 @@
...
+
+    section = parser["z80count"]
+    res = {}
+    for opt in schema:
+        v = section.get(opt.config_name)
+        try:
+            v = opt.type(v)
+        except (ValueError, TypeError):
+            perror(
+                "Error parsing config value for '%s'. Using default.",
```
Comment (reidrac) [inline z80count/z80count.py:126]: Same, please provide some top on what went wrong and potentially how to fix it.

File: z80count/z80count.py [hunk 3]
```diff
@@ -34,6 +37,180 @@
...
+    # $XDG_CONFIG_HOME is either not set or empty, a default equal to
+    # $HOME/.config should be used.
+    #
+    # https://specifications.freedesktop.org/basedir-spec/latest/ar01s03.html
+
+    xdg_config_home = os.environ.get("XDG_CONFIG_HOME")
+    if xdg_config_home is None:
+        xdg_config_home = os.path.join(home_dir, ".config")
+
+    candidate = os.path.join(xdg_config_home, "z80count")
```
Comment (reidrac) [inline z80count/z80count.py]: Use a constant for the configuration file; I prefer it is `z80countrc`.

File: z80count/z80count.py [hunk 4]
```diff
@@ -34,6 +37,180 @@
...
+    if os.path.isfile(candidate):
+        return candidate
+
+    candidate = os.path.join(home_dir, ".z80count")
```
Comment (reidrac) [inline z80count/z80count.py]: Same, use `"." + DEF_CONF` or similar.

File: z80count/z80count.py [hunk 5]
```diff
@@ -34,6 +37,180 @@
...
+
+    return None
+
+
+def parse_command_line():
+    parser = argparse.ArgumentParser(
+        description='Z80 Cycle Count',
+        epilog="Copyright (C) 2019 Juan J Martinez <jjm@usebox.net>")
+
+    # NOTE: parser arguments *must* provide a default equal to None in
```
Comment (reidrac) [inline z80count/z80count.py]: You could have used as default the value from the configuration file (that would be default if not conf file was provided).

File: README.md [hunk 6]
```diff
@@ -152,6 +152,38 @@ Where A, B, T0 and T1 are:
...
+recognized options are:
+
+- `column`: integer
+- `debug`: boolean
+- `subtotals`: boolean
+- `tab width`: int
+- `update`: boolean
+- `use tabs`: boolean
+
+Example:
```
Comment (reidrac) [inline README.md:167]: I would include a `z80countrc` as example with the defaults commented. It is easier to copy the file and uncomment/change whatever, helping the README to not have too much information (that perhaps should go into a man page).

File: README.md [hunk 7]
```diff
@@ -152,6 +152,38 @@ Where A, B, T0 and T1 are:
  - T0 is the subtotal when the conditional is not met.
  - T1 is the subtotal when the conditional is met.
 
+## Config file
+
+`z80count` will look for a config file in the following places, in order:
+
+- the file given in the environment variable `Z80COUNT_RC`.
+
+- a file `z80count` in the directory given in the environment variable
```
Comment (reidrac) [inline README.md]: I prefer this to be `z80countrc`.

**Review by reidrac — COMMENTED** [review]
A bit cumbersome for such a small tool, but I don't mind the change.

There few things I would improve, specially regarding error reporting (considering this change adds configuration file support, sounds like an important thing).

**Comment by patxoca** [comment]
I've done the requested changes. Anyway, I've been thinking that having approximately 40% of the code for managing the configuration makes little sense.

If you're OK with that I'll withdraw the PR. In the future, if a real need arises, we'll talk again.

**Comment by reidrac** [comment]
I agree it is a bit over-engineered, but I'm keeping it. Thanks!

**Comment by patxoca** [comment]
A minor detail that does not deserve an issue, in the readme, in the config file example, please remove the final period for `keep cycles` comment. I've missed that one.

**Pull request merged**